sdk/communication/azure-communication-chat/src/main/java/com/azure/android/communication/chat/ChatAsyncClient.java

```diff
@@ -170,6 +176,14 @@ public Builder credentialInterceptor(Interceptor credentialInterceptor) {
             return this;
         }
 
+        public Builder userAgentInterceptor(Context context) {
```

Comment:
This should not be a public method on the builder. The builder should always create and insert all policies into the pipeline on behalf of the user - in most cases the user is not expected to understand what a "policy" is. Any policy that requires user-provided data should accept that data via a builder method, rather than accepting a policy instance. This also goes for things like `credentialInterceptor`, you should accept various 'credential' objects rather than an Interceptor.

My suggestion is to accept a Context object in the builder's constructor, and use it to create a UserAgentInterceptor which you then assign to a field. When the `build()` method is called, insert the UserAgentInterceptor in the appropriate place in the pipeline. See an example here: https://github.com/Azure/azure-sdk-for-android/blob/master/sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/StorageBlobClient.java#L791

Reply:
For testing purposes you can also create a package-private constructor for the builder that accepts whatever you want (e.g. one or multiple interceptors, or Provider instances which is our abstraction around the Context - see here: https://github.com/Azure/azure-sdk-for-android/tree/master/sdk/core/azure-core/src/main/java/com/azure/android/core/provider and an example of them being used here: https://github.com/Azure/azure-sdk-for-android/blob/master/sdk/core/azure-core/src/test/java/com/azure/android/core/http/interceptor/UserAgentInterceptorTest.java#L289